domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+		}
+		if !blockDevicesChanged(existing, devices) {
+			return nil
+		}
+
+		if err := updateBlockDevices(ctx, tx, machineUUID, devices...); err != nil {
+			return errors.Annotatef(err, "updating block devices on machine %q (%s)", machineId, machineUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func updateBlockDevices(ctx context.Context, tx *sqlair.TX, machineUUID string, devices ...blockdevice.BlockDevice) error {
+	fsTypeQuery := `SELECT * AS &FilesystemType.* FROM filesystem_type`
+	fsTypeStmt, err := sqlair.Prepare(fsTypeQuery, FilesystemType{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var fsTypes []FilesystemType
+	if err := tx.Query(ctx, fsTypeStmt).GetAll(&fsTypes); err != nil {
+		return errors.Trace(err)
+	}
+	fsTypeByName := make(map[string]int)
+	for _, fsType := range fsTypes {
+		fsTypeByName[fsType.Name] = fsType.ID
+	}
+
+	if err := removeMachineBlockDevices(ctx, tx, machineUUID); err != nil {
+		return errors.Annotatef(err, "removing existing block devices for machine %q", machineUUID)
+	}
+
+	if len(devices) == 0 {
+		return nil
+	}
+
+	insertQuery := `
+INSERT INTO block_device (uuid, name, label, device_uuid, hardware_id, wwn, bus_address, serial_id, mount_point, size, filesystem_type_id, in_use)
+VALUES (
+    $BlockDevice.uuid,
+    $BlockDevice.name,
+    $BlockDevice.label,
+    $BlockDevice.device_uuid,
+    $BlockDevice.hardware_id,
+    $BlockDevice.wwn,
+    $BlockDevice.bus_address,
+    $BlockDevice.serial_id,
+    $BlockDevice.mount_point,
+    $BlockDevice.size,
+    $BlockDevice.filesystem_type_id,
+    $BlockDevice.in_use
+)
+`
+	insertStmt, err := sqlair.Prepare(insertQuery, BlockDevice{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	blockDevicesByUUID := make(map[utils.UUID]blockdevice.BlockDevice, len(devices))
+	for _, bd := range devices {
+		fsTypeID, ok := fsTypeByName[bd.FilesystemType]
+		if !ok {
+			return errors.NotValidf("filesystem type %q for block device %q", bd.FilesystemType, bd.DeviceName)
+		}
+		id, err := utils.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		blockDevicesByUUID[id] = bd
+		dbBlockDevice := BlockDevice{
+			ID:             id.String(),
+			DeviceName:     bd.DeviceName,
+			Label:          bd.Label,
+			DeviceUUID:     bd.UUID,
+			HardwareId:     bd.HardwareId,
+			WWN:            bd.WWN,
+			BusAddress:     bd.BusAddress,
+			SerialId:       bd.SerialId,
+			MountPoint:     bd.MountPoint,
+			Size:           bd.Size,
+			FilesystemType: fsTypeID,
+			InUse:          bd.InUse,
+		}
+		if err := tx.Query(ctx, insertStmt, dbBlockDevice).Run(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	insertLinkQuery := `
+INSERT INTO block_device_link_device (block_device_uuid, name)
+VALUES (
+    $DeviceLink.block_device_uuid,
+    $DeviceLink.name
+)
+`
+	insertStmt, err = sqlair.Prepare(insertLinkQuery, DeviceLink{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for uuid, bd := range blockDevicesByUUID {
```

Comment:
Small life hack if you wanted but we could do this grunt work in the for loop above as well. It would reduce the carbon foot print of Juju 😁.

Just a nit. Up to you.

Reply:
Can do. I find it a bit easier to grok if things are kept separate, but I did the change.